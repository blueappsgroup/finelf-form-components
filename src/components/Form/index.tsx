import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
