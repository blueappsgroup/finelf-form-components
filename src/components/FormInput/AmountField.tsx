import React, { FC } from 'react'
import { Field } from 'formik'

import { RangeFieldProps } from '../../types'
import { BaseRangeField } from './base'

const AmountField: FC<RangeFieldProps> = ({
  id,
  label = 'Kwota pożyczki',
  name,
  required = false,
  showError = true,
  showMinMax = true,
  min = 300,
  max = 10000,
  step = 100,
  value = 2000,
  suffix = 'PLN',
  unit = 'zł',
}) => (
  <Field
    required={required}
    showError={showError}
    showMinMax={showMinMax}
    id={id}
    name={name}
    component={BaseRangeField}
    label={label}
    type="text"
    min={min}
    max={max}
    step={step}
    value={value}
    suffix={suffix}
    unit={unit}
  />
)

AmountField.defaultProps = {
  name: 'amount',
}

export default AmountField
