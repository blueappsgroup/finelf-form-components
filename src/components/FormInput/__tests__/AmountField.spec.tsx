import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import AmountField from '../AmountField'

describe('<AmountField />', () => {
  const onSubmit = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const setupWrapper: (config: any) => any = ({
    formId = 'testForm',
    inputId = 'amountField',
    inputName = 'name',
    required = false,
    ...rest
  }) => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <AmountField
          id={inputId}
          name={inputName}
          required={required}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const input = container.querySelector(`[name="${inputName}"]`)

    return {
      input,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { input } = setupWrapper({})

    expect(input).toBeTruthy()
    expect(input).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes input field', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: '400' } })
    })

    expect(input.value).toBe('400')
  })

  it('input field onBlur save value to sessionStorage', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: '400' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: '400' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm')).name
    ).toBe(400)
    expect(input.value).toBe('400')
  })
})
