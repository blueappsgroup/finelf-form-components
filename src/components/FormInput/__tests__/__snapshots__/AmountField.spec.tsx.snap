// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmountField /> matches snapshot 1`] = `
<input
  class="sc-bkzZxe hpQztk"
  id="amountField"
  label="Kwota pożyczki"
  max="10000"
  min="300"
  name="name"
  step="100"
  type="text"
  value="2000"
/>
`;

exports[`<AmountField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <div
        class="sc-crrsfI ieBApp"
      >
        <label
          for="name"
        >
          Test
        </label>
        <div
          class="sc-fFubgz fJyOdk"
        >
          <input
            class="sc-bkzZxe hpQztk"
            id="amountField"
            label="Test"
            max="10000"
            min="300"
            name="name"
            step="100"
            type="text"
            value="2000"
          />
          <span
            class="sc-bBXqnf bwzWoF"
          >
            PLN
          </span>
        </div>
      </div>
      <div
        class="sc-lmoMRL hIjsuX"
      >
        <div
          aria-orientation="horizontal"
          aria-valuemax="10000"
          aria-valuemin="300"
          aria-valuenow="2000"
          class="rangeslider rangeslider-horizontal"
        >
          <div
            class="rangeslider__fill"
            style="width: 0px;"
          />
          <div
            class="rangeslider__handle"
            style="left: 0px;"
            tabindex="0"
          >
            <div
              class="rangeslider__handle-label"
            />
          </div>
          <ul
            class="rangeslider__labels"
          />
        </div>
      </div>
      <div
        class="sc-crrsfI ieBApp"
      >
        <span
          class="sc-kfzAmx jbFYGv"
        >
          Min. 
          300
           
          zł
        </span>
        <span
          class="sc-kfzAmx jbFYGv"
        >
          Max. 
          10000
           
          zł
        </span>
      </div>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;
